### PR TITLE
Add placement support

### PR DESCRIPTION
**Pull request by shepmaster**
THIS SHOULD NOT BE MERGED YET.

This is a first-crack at adding (unstable) placement support. This definitely needs to be changed to be behind a feature flag, and I'd like to think about how this could be better tested.

I'd especially like to have some amount of performance testing, as that should be the point of the placement syntax.

I'm mostly opening this to start a dialog about the implementation and how you'd like to see this get added.

### CODE REVIEW

**Comment by SimonSapin** [comment]
I’m not really using this library anymore. Are you interested in maintaining it?

**Comment by shepmaster** [comment]
> I’m not really using this library anymore. 

That's a shame. Did you find a better alternative?

> Are you interested in maintaining it?

I could certainly take up that mantle; it doesn't look like there's toooo much traffic.

**Comment by SimonSapin** [comment]
I ended up using `Rc`, sacrificing some performance for a more convenient API. Discussion is at https://github.com/SimonSapin/rust-forest, and the actual library at https://github.com/SimonSapin/kuchiki.

File: lib.rs [hunk 1]
```diff
@@ -145,3 +150,131 @@ fn it_works() {
...
+    impl Drop for Monster {
+        fn drop(&mut self) {
+            assert_eq!(self.health, 42);
+        }
+    }
+
+    // This test is extremely odd. We really want to make sure that
+    // the Monster destructor is called exactly once, but we can't
+    // have any known state. This is a case where a mutable global
+    // variable would be quite useful.
```
Comment (huonw) [inline lib.rs]: You can use a `static mut` + `unsafe`, or `static` with an atomic:

```rust
use std::sync::atomic::{AtomicUsize, Ordering, ATOMIC_USIZE_INIT};

static COUNT: AtomicUsize = ATOMIC_USIZE_INIT;

impl Drop for Monster {
    fn drop(&mut self) {
        COUNT.fetch_add(1, Ordering::SeqCst);
    }
}

assert!(thread::spawn(|| {
    let arena = Arena::new();
    // ...
}).join().is_err());

assert_eq!(COUNT.load(Ordering::SeqCst), 1);
```

**Comment by jonastepe** [comment]
If shepmaster isn't interested anymore, I would be happy to maintain this library. Though I am still learning Rust, I am well beyond a beginner. This is a nice and small project. For me, it is an ideal starting point towards a more active contribution to the Rust ecosystem.

**Comment by SimonSapin** [comment]
Hi @jonastepe. I’ve given you write access to this repository, but crates.io didn’t know about your username when I tried to do it for https://crates.io/crates/typed-arena. Have you logged in there?

**Comment by jonastepe** [comment]
@SimonSapin it should be possible now. I've logged in on crates.io. Thank you!

**Comment by SimonSapin** [comment]
Done :)

**Comment by jonastepe** [comment]
I will start maintaining then ;)
Thanks again for the opportunity.

**Comment by pczarn** [comment]
Can this be rebased and merged now? @jonastepe

**Comment by shepmaster** [comment]
@pczarn I'll try to rebase this evening (Eastern)

**Comment by shepmaster** [comment]
@pczarn updated. It only works on nightly, but I'd like maintainer feedback on how it should be merged (feature flags, waiting until it's stable, etc.).

[Some performance numbers](https://gist.github.com/shepmaster/6c0da90d099e80e01464) from my [performance branch](https://github.com/shepmaster/rust-typed-arena/tree/placement-benchmarks).

**Pull request closed without merging**